src/weights.jl

```diff
@@ -27,6 +27,12 @@ isempty(wv::AbstractWeights) = isempty(wv.values)
 Base.getindex(wv::AbstractWeights, i) = getindex(wv.values, i)
 Base.size(wv::AbstractWeights) = size(wv.values)
 
+function Base.isequal(w₁::W, w₂::W) where {W <: AbstractWeights}
+    isequal(sum(w₁), sum(w₂)) && isequal(values(w₁), values(w₂))
+end
+
+Base.isequal(w₁::AbstractWeights, w₂::AbstractWeights) = false
```

Comment:
I'm guessing this is the fall back condition where the `typeof(w₁) != typeof(w₂)`, but it might we worth including a comment.

Reply:
Is it really needed though? The fallback to `===` should be enough.

Reply:
I've just realized that this two-method implementation is equivalent to `x === y`. It returns `false` if `x = fweights([1; 2; 3])` and `y = fweights([1.0; 2.0; 3.0])`, whereas `isequal([1; 2; 3], [1.0; 2.0; 3.0])` returns true. It is stricter than I intended – namely, distinguishing between the different weight types (`FrequencyWeights`, `ProbabilityWeights`). I don't know if we want to return `false` because a difference in parameters. I'm trying to figure out a way around it (other than brute force).

Reply:
Using `isequal` as you do is fine here, indeed we want `1` and `1.0` to be considered as equal. What I meant is that you should be able to get rid of the `= false` method.

Reply:
But then `isequal(x, y) = true` if `x = fweights([1; 2; 3])` and `y = pweights([1; 2; 3])`

Reply:
Not if you define only `Base.isequal(w₁::W, w₂::W) where {W <: AbstractWeights}` as you do above.

Reply:
It does. As a matter of fact, `isequal(x, y) = true` even if `x = FrequencyWeights([1; 2; 3], 6)` and `y = ProbabilityWeights([1; 2; 3], 0)` when we only define the first method.

Here's my guess. Suppose that we only define the first method. When `x` and `y` have the same type and the same parameters, the new method is called to check the values and the sum. When `x ` and `y` have different types or different parameters, the default method is called. Because `AbstractWeights` are subtypes of `AbstractVector`, the fallback is `isequal(x::Array, y::Array)`, which only cares about the `values` field.

Reply:
Ah, right, then we indeed need the method. Probably worth adding a comment to explain why it's there.

Reply:
Here's a proposition:

    for w in [AnalyticWeights; FrequencyWeights; ProbabilityWeights; Weights]
        eval(quote
            Base.isequal(x::$w, y::$w) = isequal(x.sum, y.sum) && isequal(x.values, y.values)
            Base.:(==)(x::$w, y::$w)   = (x.sum == y.sum) && (x.values == y.values)
        end)
    end

    Base.isequal(x::AbstractWeights, y::AbstractWeights) = false
    Base.:(==)(x::AbstractWeights, y::AbstractWeights)   = false

Under this implementation, parameters aren't a problem: `isequal(x, y) = true` if `x = fweights([1])` and `y = fweights([1.0])`, but `isequal(x, y) = false` if `y = pweights[1.0]`.

Does it seem reasonable?

Reply:
Yeah, but I don't think you need the loop. Isn't `Base.isequal(w₁::W, w₂::W) where {W <: AbstractWeights}` enough, rather than defining a separate function for each type?

Reply:
Under that implementation, `isequal(x, y) = false` if `x = weights([1])` and `y = weights([1.0])`.* I thought we didn't want that behavior.

* Reason being: `x` is `Weights{Int64, Int64, Vector{Int64}}` and `y` is `Weights{Float64, Float64, Vector{Float64}}`, so `isequal(x, y)` calls the fallback.

Reply:
Damn, you're right. I can't believe I repeatedly got this wrong so many times...

Reply:
This seems like something we should be able to express with an appropriately parameterized method.

Reply:
Maybe... But which one? :-p

Reply:
How about:
```julia
function Base.isequal(x::W, y::W) where {W <: AbstractWeights}
    isequal(sum(x), sum(y)) && isequal(values(x), values(y))
end

function Base.isequal(x::AbstractWeights, y::AbstractWeights)
    Base.datatype_name(typeof(x)) == Base.datatype_name(typeof(y)) || throw(MethodError(Base.isequal, (x, y)))
    isequal(sum(x), sum(y)) && isequal(values(x), values(y))
end
```

Reply:
That would also work if someone wants to define their own custom weights type... for some reason :)

Reply:
Yeah. But using the unexported `Base.datatype` name isn't terribly clean. And the generic definition could even be incorrect if somebody defines an `AbstractWeights` type with a type parameter specifying the kind of weights (i.e. types with different parameters should be considered as different).

Reply:
Both implementations work. I leave the decision to smarter people than me. :)